[rpc] Support netty metrics

SteNicholas opened the pull request:
### Purpose

Linked issue: close #135

Support netty metrics including:

- `netty`
  - `usedDirectMemory`
  - `numDirectArenas`
  - `numAllocationsPerSecond`
  - `numHugeAllocationsPerSecond`

### Tests

`NettyMetricsTest`

### API and Format

No.

### Documentation

`monitor-metrics.md` adds the `usedDirectMemory`, `numDirectArenas`, `numAllocationsPerSecond` and `numHugeAllocationsPerSecond` metrics.

SteNicholas commented:
Ping @wuchong.

wuchong commented on fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/NettyMemoryMetrics.java:
```diff
@@ -0,0 +1,169 @@
+            MetricGroup metricGroup,
+            PooledByteBufAllocator pooledAllocator,
+            boolean verboseMetricsEnabled) {
+        this.metricGroup = metricGroup.addGroup(NETTY_MEMORY_METRIC_GROUP);
```
Use "netty" metric group, so that we can add non-memory metrics into it in the future. And rename the class to `NettyMetrics` as well.

wuchong commented on fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/NettyMemoryMetrics.java:
```diff
@@ -0,0 +1,169 @@
+                        "numActiveBytes"));
+    }
+
+    public NettyMemoryMetrics(
```
We don't need to instantiate the class, just expose a `registerNettyMetrics(..)` util method is enough.

wuchong commented on fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/NettyMemoryMetrics.java:
```diff
@@ -0,0 +1,169 @@
+    public static final String NETTY_MEMORY_NUM_THREAD_LOCAL_CACHES = "numThreadLocalCaches";
+    public static final String NETTY_MEMORY_CHUNK_SIZE = "chunkSize";
+    public static final String NETTY_MEMORY_HEAP_ARENA = "heapArena";
+    public static final String NETTY_MEMORY_DIRECT_ARENA = "directArena";
```
Let's move all the metric names to `MetricNames` where we maintain all the metric names centralized.

wuchong reviewed (commented):
Thanks @SteNicholas for the contribution. 

The implementation is great. However, I think most of the registered metrics will never be used. We can reduce the metrics to simplify the implementation and costs, necessary metrics can be introduced when needed in the future. 

Some background:
1. Fluss never allocates heap buffers via Netty, so only direct buffers need to be monitored
2. Fluss may allocate huge buffers for high-throughput log producing scenarios, and huge allocation is time-consuming, so we should monitor huge allocations.

So the suggested metrics to monitor: 

- `usedDirectMemory`: Gauge, direct memory size used by netty
- `numDirectArenas`: Gauge, number of direct arenas in netty
- `allocationsPerSecond`: Meter, the number of allocations every second
- `hugeAllocationsPerSecond`: Meter, the number of huge allocations every second

We don't need to monitor detailed arenas-level metrics, this can simplify the code by removing Java reflection and reduce a lot of metrics. 

Btw, we also need to add documentation for the new metrics in the page: https://alibaba.github.io/fluss-docs/docs/maintenance/monitor-metrics/#threads . We can add a specific section for "Netty", the scope can be `
coordinator/tabletserver/client`.

SteNicholas commented:
@wuchong, thanks for detailed review and suggestions. I have addressed comments in https://github.com/alibaba/fluss/pull/407/commits/d54880d1adde170b46f8a9030c1dbdae78bb9536. PTAL.

wuchong commented on fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/NettyMetrics.java:
```diff
@@ -0,0 +1,86 @@
+                new MeterView(
+                        new AllocationsCounter(
+                                () ->
+                                        pooledAllocator.directArenas().stream()
```
This is deprecated. Use `pooledAllocatorMetric.directArenas()` instead.

wuchong commented on fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/NettyMetrics.java:
```diff
@@ -0,0 +1,86 @@
+        nettyMetricGroup.meter(
+                MetricNames.NETTY_NUM_ALLOCATIONS_PER_SECONDS,
+                new MeterView(
+                        new AllocationsCounter(
```
We don't need to introduce `AllocationsCounter`, here can be 

```java
        nettyMetricGroup.meter(
                MetricNames.NETTY_NUM_ALLOCATIONS_PER_SECONDS,
                new MeterView(
                        () ->
                                pooledAllocatorMetric.directArenas().stream()
                                        .mapToLong(PoolArenaMetric::numAllocations)
                                        .sum()));
```

wuchong commented on fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/NettyMetrics.java:
```diff
@@ -0,0 +1,86 @@
+                                                .sum())));
+        nettyMetricGroup.meter(
+                MetricNames.NETTY_NUM_HUGE_ALLOCATIONS_PER_SECONDS,
+                new MeterView(
+                        new AllocationsCounter(
+                                () ->
+                                        pooledAllocator.directArenas().stream()
+                                                .mapToLong(PoolArenaMetric::numHugeAllocations)
```
ditto.

wuchong reviewed (approved):
LGTM.
